File: RVersion/R/effect_sets.R
#' @import data.table
#' @importFrom stats as.formula coef model.matrix na.omit
NULL

#' Build the full set of bounds for effect pairs
#'
#' Build the full set of bounds for effect pairs based on partial identification bounds.
#' @include helpers.R
#' @param bounds_data A data frame containing the partial identification bounds for each individual setting's effect, from a function like \code{ovb_bounds_by_setting}. Must be a \code{data.frame} with a column for \code{setting} and columns for \code{original_estimate}, \code{adjusted_estimate}, \code{lower_plausible_bound}, and \code{upper_plausible_bound}. Other columns are acceptable.
#' @param rho_bounds A matrix with the upper bound of rho_ij in the ith row and jth column, and the lower bound in the ith column and jth row. Trace values will be ignored. If all lower triangle values are 0, they will be set to the inverse of the corresponding upper bound. Alternately, provide a single real value to bound all pairs at that same value and its inverse. Set a bound to 0 to ignore it (set no restriction).
#' @param settings_order A vector with the values of the settings in the order they are used in \code{rho_bounds}. Defaults to the order of settings in \code{bounds_data}.
#' @param ... Additional arguments (not currently used).
#'
#' @examples
#'
#' library(fixest)
#' library(hetset)
#' data("close_college")
#'
#' bounds = ovb_bounds_by_setting(close_college,
#'                                setting = 'blacksouth',
#'                                treatment = 'educ',
#'                                model = 'feols(lwage ~ educ + smsa + married + exper, data = data)',
#'                                benchmark_covariates = c('smsa','married','exper'),
#'                                kd = .1)
#'
#' # Put a 1.5 bound on Black vs Black & South and Neither vs South,
#' # 2 on Neither vs. black and South vs. Black & South
#' # and no restriction on Neither vs Black & South and on Black vs. South (setting to 0)
#' # Setting all the lower triangle values of 0
#' # will make them automatically the inverse of the upper triangle values
#' boundmat = matrix(c(0, 2, 1.5, 0,
#'                     0, 0, 0, 1.5,
#'                     0, 0, 0, 2,
#'                     0, 0, 0,0), nrow = 4, byrow = TRUE)
#' # Get the partial identification bounds
#' pi_bounds = build_bounds(bounds, boundmat,
#'                             settings_order = c('Neither','Black','South','Black and South'))
#'
#' @export

build_bounds = function(bounds_data,
                        rho_bounds,
                        settings_order = bounds_data$setting,
                        ...) {

  data.table::setDT(bounds_data)

  # ensure that settings_order contains all the settings and no more
  settings_not_in_order = setdiff(bounds_data$setting, settings_order)
  order_not_in_settings = setdiff(settings_order, bounds_data$setting)
  if (length(settings_not_in_order) > 0) {
    stop(paste0("The following settings are in bounds_data but not in settings_order: ", paste0(settings_not_in_order, collapse = ", ")))
  }
  if (length(order_not_in_settings) > 0) {
    stop(paste0("The following settings are in settings_order but not in bounds_data: ", paste0(order_not_in_settings, collapse = ", ")))
  }

  if (!is.matrix(rho_bounds)) {
    # Ensure it's just a single real value
    if (!is.numeric(rho_bounds) | length(rho_bounds) != 1) {
      stop('If rho_bounds is not a matrix it must be a single real value.')
    }
    rbvalue = rho_bounds
    rho_bounds = matrix(rep(rbvalue, length(settings_order)^2), ncol = length(settings_order))
    rho_bounds[lower.tri(rho_bounds)] = 1/rbvalue
  }

  # rho_bounds should be square with rows and cols equal to the length of settings_order
  if (nrow(rho_bounds) != ncol(rho_bounds) | nrow(rho_bounds) != length(settings_order)) {
    stop("rho_bounds must be a square matrix with dimensions equal to the length of settings_order.")
  }

  univariate_bounds_list = list()
  setting = NULL
  for (i in 1:length(settings_order)) {
    s = settings_order[i]
    bounds_s = bounds_data[setting == s]
    univariate_bounds_list[[s]] = list(
      estimate = bounds_s$original_estimate,
      lower = bounds_s$lower_plausible_bound,
      upper = bounds_s$upper_plausible_bound,
      lower_nu = bounds_s$original_estimate - bounds_s$upper_plausible_bound,
      upper_nu = bounds_s$original_estimate - bounds_s$lower_plausible_bound
    )
  }

  # Check if all sub-trace elements of rho_bounds are 0
  all_subtrace_zero = all(rho_bounds[lower.tri(rho_bounds)] == 0)

  paired_bounds_matrix = matrix(list(), nrow = length(settings_order), ncol = length(settings_order))
  for (i in 1:(length(settings_order)-1)) {
    for (j in (i+1):length(settings_order)) {
      s_i = settings_order[i]
      s_j = settings_order[j]
      if (rho_bounds[i,j] == 0) {
        pair_bounds = c(-Inf, Inf)
      } else {
        pair_bounds = c_bounds(
          b_lower = univariate_bounds_list[[s_j]]$lower_nu,
          b_upper = univariate_bounds_list[[s_j]]$upper_nu,
          rho_upper = rho_bounds[i,j],
          rho_lower = ifelse(all_subtrace_zero, 1/rho_bounds[i,j], rho_bounds[j,i]),
          return_full_set = FALSE
        )
      }
      paired_bounds_matrix[i, j] = pair_bounds[2]
      paired_bounds_matrix[j, i] = pair_bounds[1]
    }
  }
  colnames(paired_bounds_matrix) = settings_order
  rownames(paired_bounds_matrix) = settings_order
  return(list(
    univariate_bounds = univariate_bounds_list,
    paired_bounds = paired_bounds_matrix
  ))
}


#' Check for the existence of an identified set satisfying all bounds
#'
#' @param full_bounds_set A full set of univariate and paired bounds produced by \code{build_bounds}
#' @param check_only A vector of setting names to check only those settings. If \code{NULL} (default) checks all settings. This option is used to try to figure out which setting combinations might be leading to no identified set existing.
#' @param return_one_solution Set to \code{TRUE} to return a set of parameters that satisfy all bounds. \emph{IT IS EXTREMELY IMPORTANT TO NOTE} that this is not a "best" solution or anything like that. It is simply one possible solution.
#'
#' @examples
#' library(fixest)
#' library(hetset)
#' data("close_college")
#'
#' bounds = ovb_bounds_by_setting(close_college,
#'                                setting = 'blacksouth',
#'                                treatment = 'educ',
#'                                model = 'feols(lwage ~ educ + smsa + married + exper, data = data)',
#'                                benchmark_covariates = c('smsa','married','exper'),
#'                                kd = .1)
#'
#' # Put a 1.5 bound on Black vs Black & South and Neither vs South,
#' # 2 on Neither vs. black and South vs. Black & South
#' # and no restriction on Neither vs Black & South and on Black vs. South (setting to 0)
#' # Setting all the lower triangle values of 0
#' # will make them automatically the inverse of the upper triangle values
#' boundmat = matrix(c(0, 2, 1.5, 0,
#'                     0, 0, 0, 1.5,
#'                     0, 0, 0, 2,
#'                     0, 0, 0,0), nrow = 4, byrow = TRUE)
#' # Get the partial identification bounds
#' boundset = build_bounds(bounds, boundmat,
#'                         settings_order = c('Neither','Black','South','Black and South'))
#'
#'
#' # Check if there exists a set of effect parameters that satisfy all bounds.
#' identified_set_exists(boundset)
#' # Get one example of a set of parameters in the identified set.
#' identified_set_exists(boundset, return_one_solution = TRUE)
#'
#' @export
identified_set_exists = function(full_bounds_set, check_only = NULL, return_one_solution = FALSE) {
  get_setup = set_up_solver(full_bounds_set, check_only)
  rst = get_setup$rst
  direc = get_setup$direc
  rhs = get_setup$rhs

  numsets = ncol(rst)

  slv = Rglpk::Rglpk_solve_LP(
    obj = rep(0, numsets),
    mat = rst,
    dir = direc,
    rhs = rhs
  )

  if (slv$status == 0) {
    if (return_one_solution) {
      solution = slv$solution
      settingnames = names(full_bounds_set[[1]])
      if (!is.null(check_only)) {
        settingnames = check_only
      }
      names(solution) = settingnames
      return(solution)
    } else {
      return(TRUE)
    }
  } else {
    return(FALSE)
  }
}


#' Univariate Bounds Table
#'
#' Creates a table of bounds for the effect in each setting.
#'
#' The table shows the widest possible bounds consistent with the assumptions for each setting one at a time.
#'
#' Note that because this is a highly multidimensional problem, the actual identified set of plausible estimates is somewhat narrower.
#'
#' For example, if in setting A the effect must be between 0 and 1, in setting B it must be between 1 and 2, and the difference between them must be no greater than .5, then the univariate set for A will be .5 to 1 and the set for B will be 1 to 1.5, ignoring that the joint set (.5 for A, 1.5 for B) is not in the identified set.
#'
#' To check whether a specific combination of values is in the identified set, use the function \code{valid_solution_check}.
#'
#' @param full_bounds_set A full set of univariate and paired bounds produced by \code{build_bounds}.
#' @param pin_values A named list of settings and values to pin estimates in those settings to. Values can be numeric, or can be \code{'lower'}, \code{'upper'}, or \code{'original'} for the univariate lower-bound, upper bound, or original estimate, respectively. Useful for exploring the multivariate nature of the bounds.
#' @param ... Additional arguments (not currently used).
#'
#' @examples
#' library(fixest)
#' library(hetset)
#' data("close_college")
#' bounds = ovb_bounds_by_setting(close_college,
#'                                setting = 'blacksouth',
#'                                treatment = 'educ',
#'                                model = 'feols(lwage ~ educ + smsa + married + exper, data = data)',
#'                                benchmark_covariates = c('smsa','married','exper'),
#'                                kd = .1)
#'
#' # Put a 1.5 bound on Black vs Black & South and Neither vs South,
#' # 2 on Neither vs. black and South vs. Black & South
#' # and no restriction on Neither vs Black & South and on Black vs. South (setting to 0)
#' # Setting all the lower triangle values of 0
#' # will make them automatically the inverse of the upper triangle values
#' boundmat = matrix(c(0, 2, 1.5, 0,
#'                     0, 0, 0, 1.5,
#'                     0, 0, 0, 2,
#'                     0, 0, 0,0), nrow = 4, byrow = TRUE)
#' # Get the partial identification bounds
#' boundset = build_bounds(bounds, boundmat,
#'                         settings_order = c('Neither','Black','South','Black and South'))
#'
#' # Get the univariate bounds table
#' univariate_bounds_table(boundset)
#' # Note in this example that the bounds are only very
#' # slightly tighter than if we hadn't worried about cross-setting
#' # relationships (for Black).
#'
#' @export
univariate_bounds_table = function(full_bounds_set, pin_values = NULL, ...) {
  setting_names = names(full_bounds_set[[1]])

  original_bounds_set = copy(full_bounds_set)

  if (!is.null(pin_values)) {
    for (s in names(pin_values)) {
      set_value = pin_values[[s]]
      if (!is.numeric(set_value)) {
        if (set_value == 'lower') {
          set_value = full_bounds_set[[1]][[s]]$lower
        } else if (set_value == 'upper') {
          set_value = full_bounds_set[[1]][[s]]$upper
        } else if (set_value == 'original') {
          set_value = full_bounds_set[[1]][[s]]$estimate
        } else {
          stop(paste0("pin_values for setting ", s, " must be numeric, 'lower', 'upper', or 'original'."))
        }
      }
      # Modify the full_bounds_set to pin this value
      full_bounds_set[[1]][[s]]$lower = set_value
      full_bounds_set[[1]][[s]]$upper = set_value
    }
  }

  get_setup = set_up_solver(full_bounds_set, check_only = NULL)
  rst = get_setup$rst
  direc = get_setup$direc
  rhs = get_setup$rhs

  numsets = length(setting_names)

  full_bounds_table = data.table::data.table()

  for (s in 1:numsets) {
    mincol = rep(0, numsets)
    mincol[s] = 1
    slv_low = Rglpk::Rglpk_solve_LP(
      obj = mincol,
      mat = rst,
      dir = direc,
      rhs = rhs
    )$solution[s]
    maxcol = rep(0, numsets)
    maxcol[s] = -1
    slv_up = Rglpk::Rglpk_solve_LP(
      obj = maxcol,
      mat = rst,
      dir = direc,
      rhs = rhs
    )$solution[s]
    full_bounds_table = rbind(
      full_bounds_table,
      data.table::data.table(
        setting = setting_names[s],
        estimate = full_bounds_set[[1]][[s]]$estimate,
        lower_bound = slv_low,
        upper_bound = slv_up,
        lower_nu_bound = full_bounds_set[[1]][[s]]$estimate - slv_up,
        upper_nu_bound = full_bounds_set[[1]][[s]]$estimate - slv_low,
        original_lower_bound = original_bounds_set[[1]][[s]]$lower,
        original_upper_bound = original_bounds_set[[1]][[s]]$upper
      )
    )
  }

  estimate = NULL
  upper_nu_bound = NULL
  lower_nu_bound = NULL
  lower_bound = NULL
  upper_bound = NULL
  full_bounds_table[, lower_bound := estimate - upper_nu_bound]
  full_bounds_table[, upper_bound := estimate - lower_nu_bound]

  rst_dt = data.table::data.table(
    setting = setting_names,
    lower_bound = sapply(setting_names, function(s) full_bounds_set[[1]][[s]]$lower),
    upper_bound = sapply(setting_names, function(s) full_bounds_set[[1]][[s]]$upper)
  )
  return(full_bounds_table)
}

File: Python/hetset/effect_sets.py
"""
effect_sets.py
--------------
Construction and analysis of the partial identification bounds set.

Provides the main workflow functions for building the full set of
univariate and pairwise bounds (:func:`build_bounds`), checking whether
an identified set exists (:func:`identified_set_exists`), and summarising
the widest per-setting bounds consistent with all assumptions
(:func:`univariate_bounds_table`).
"""

import copy
import numpy as np
import pandas as pd

from .helpers import c_bounds, set_up_solver, _linprog_solve


def build_bounds(bounds_data, rho_bounds, settings_order=None, **kwargs):
    """
    Build the Full Set of Bounds for Effect Pairs.

    Combine per-setting OVB plausibility intervals (from e.g.
    :func:`~hetset.effect_estimate_sensemakr.ovb_bounds_by_setting`) with
    cross-setting rho constraints to produce the complete bounds structure
    consumed by :func:`identified_set_exists` and
    :func:`univariate_bounds_table`.

    Parameters
    ----------
    bounds_data : pandas.DataFrame
        Per-setting partial identification bounds.  Must contain columns
        ``setting``, ``original_estimate``, ``lower_plausible_bound``, and
        ``upper_plausible_bound``.  Typically the output of
        :func:`~hetset.effect_estimate_sensemakr.ovb_bounds_by_setting`.
    rho_bounds : numpy.ndarray or float
        * **Matrix** – square, with ``len(settings_order)`` rows and
          columns.  Element ``[i, j]`` is the *upper* bound on rho for the
          pair ``(settings_order[i], settings_order[j])``; element
          ``[j, i]`` is the corresponding *lower* bound.  If all
          lower-triangle values are 0 they are automatically set to
          ``1 / upper_triangle``.  Set an element to 0 to impose no
          restriction on that pair.
        * **Scalar** – a single positive float applied as the upper bound
          to every pair; the lower bound is then ``1 / scalar``.
    settings_order : list, optional
        Values of the setting variable in the order corresponding to the
        rows/columns of ``rho_bounds``.  Defaults to the order of settings
        as they appear in ``bounds_data``.
    **kwargs
        Currently unused; accepted for forward compatibility.

    Returns
    -------
    dict
        A dictionary with two keys:

        ``'univariate_bounds'``
            dict mapping each setting name to a sub-dict with keys
            ``estimate``, ``lower``, ``upper``, ``lower_nu``,
            ``upper_nu``.

        ``'paired_bounds'``
            :class:`pandas.DataFrame` with settings as both index and
            columns.  Element ``[s_i, s_j]`` stores the *upper* bound on
            the bias-difference ``(nu_i - nu_j)``; element ``[s_j, s_i]``
            stores the *lower* bound.

    Examples
    --------
    ::

        import numpy as np
        from hetset import ovb_bounds_by_setting, build_bounds, load_close_college

        df = load_close_college()
        bounds = ovb_bounds_by_setting(
            data=df, setting='blacksouth', treatment='educ',
            model='lwage ~ educ + smsa + married + exper',
            benchmark_covariates=['smsa', 'married', 'exper'], kd=0.1,
        )

        # 4x4 rho matrix (lower triangle 0 => auto-inverted)
        boundmat = np.array([
            [0,   2,   1.5, 0  ],
            [0,   0,   0,   1.5],
            [0,   0,   0,   2  ],
            [0,   0,   0,   0  ],
        ])
        pi_bounds = build_bounds(
            bounds, boundmat,
            settings_order=['Neither', 'Black', 'South', 'Black and South'],
        )
    """
    if settings_order is None:
        settings_order = list(bounds_data['setting'])

    # --- Validate settings match -----------------------------------------
    settings_in_data = set(bounds_data['setting'])
    settings_in_order = set(settings_order)

    missing_from_order = settings_in_data - settings_in_order
    missing_from_data = settings_in_order - settings_in_data

    if missing_from_order:
        raise ValueError(
            "The following settings are in bounds_data but not in "
            f"settings_order: {', '.join(str(s) for s in missing_from_order)}"
        )
    if missing_from_data:
        raise ValueError(
            "The following settings are in settings_order but not in "
            f"bounds_data: {', '.join(str(s) for s in missing_from_data)}"
        )

    # --- Normalise rho_bounds to a matrix --------------------------------
    if not isinstance(rho_bounds, np.ndarray):
        if not np.isscalar(rho_bounds):
            raise ValueError(
                "If rho_bounds is not a matrix it must be a single real value."
            )
        rb_val = float(rho_bounds)
        n = len(settings_order)
        rho_bounds = np.full((n, n), rb_val)
        rho_bounds[np.tril_indices(n)] = 1.0 / rb_val
        np.fill_diagonal(rho_bounds, 0.0)

    n = len(settings_order)
    if rho_bounds.shape != (n, n):
        raise ValueError(
            "rho_bounds must be a square matrix with dimensions equal to "
            "the length of settings_order."
        )

    # --- Build univariate bounds -----------------------------------------
    univariate_bounds_dict = {}
    for s in settings_order:
        row = bounds_data[bounds_data['setting'] == s].iloc[0]
        univariate_bounds_dict[s] = {
            'estimate': float(row['original_estimate']),
            'lower': float(row['lower_plausible_bound']),
            'upper': float(row['upper_plausible_bound']),
            'lower_nu': float(
                row['original_estimate'] - row['upper_plausible_bound']
            ),
            'upper_nu': float(
                row['original_estimate'] - row['lower_plausible_bound']
            ),
        }

    # --- Check whether all lower-triangle values are 0 ------------------
    lower_tri_vals = rho_bounds[np.tril_indices(n, k=-1)]
    all_subtrace_zero = np.all(lower_tri_vals == 0)

    # --- Build pairwise bounds -------------------------------------------
    paired_bounds = pd.DataFrame(
        data=np.full((n, n), np.nan),
        index=settings_order,
        columns=settings_order,
    )

    for i in range(n - 1):
        for j in range(i + 1, n):
            s_i = settings_order[i]
            s_j = settings_order[j]

            if rho_bounds[i, j] == 0:
                pair_bounds = [-np.inf, np.inf]
            else:
                rho_lower_ij = (
                    1.0 / rho_bounds[i, j]
                    if all_subtrace_zero
                    else rho_bounds[j, i]
                )
                pair_bounds = c_bounds(
                    b_lower=univariate_bounds_dict[s_j]['lower_nu'],
                    b_upper=univariate_bounds_dict[s_j]['upper_nu'],
                    rho_upper=rho_bounds[i, j],
                    rho_lower=rho_lower_ij,
                    return_full_set=False,
                )

            # [i, j] = upper bound; [j, i] = lower bound
            paired_bounds.loc[s_i, s_j] = pair_bounds[1]
            paired_bounds.loc[s_j, s_i] = pair_bounds[0]

    return {
        'univariate_bounds': univariate_bounds_dict,
        'paired_bounds': paired_bounds,
    }


def identified_set_exists(full_bounds_set, check_only=None,
                          return_one_solution=False):
    """
    Check for the Existence of an Identified Set Satisfying All Bounds.

    Uses linear programming to determine whether there exists a vector of
    true treatment effects — one per setting — that simultaneously
    satisfies all univariate and pairwise bounds.

    Parameters
    ----------
    full_bounds_set : dict
        Output of :func:`build_bounds`.
    check_only : list of str, optional
        Setting names to restrict the check to.  If ``None`` (default),
        all settings are checked.  Must contain at least two settings when
        provided; useful for diagnosing which pair(s) of settings drive
        infeasibility.
    return_one_solution : bool, optional
        If ``True`` and the identified set exists, return a dict mapping
        each (checked) setting name to a feasible effect value rather than
        returning ``True``.

        .. important::
            This is **not** a "best" or "representative" solution.  It is
            simply *one* feasible point in the identified set, determined
            by the LP solver's internal path.

    Returns
    -------
    bool or dict
        ``True`` / ``False`` indicating whether the identified set is
        non-empty, or — when ``return_one_solution=True`` and the set
        exists — a dict ``{setting: feasible_value}``.

    Examples
    --------
    ::

        from hetset import (
            ovb_bounds_by_setting, build_bounds,
            identified_set_exists, load_close_college,
        )
        import numpy as np

        df = load_close_college()
        bounds = ovb_bounds_by_setting(
            data=df, setting='blacksouth', treatment='educ',
            model='lwage ~ educ + smsa + married + exper',
            benchmark_covariates=['smsa', 'married', 'exper'], kd=0.1,
        )
        boundmat = np.array([
            [0, 2, 1.5, 0],
            [0, 0, 0, 1.5],
            [0, 0, 0, 2],
            [0, 0, 0, 0],
        ])
        boundset = build_bounds(
            bounds, boundmat,
            settings_order=['Neither', 'Black', 'South', 'Black and South'],
        )

        # Check feasibility
        identified_set_exists(boundset)

        # Get one feasible solution
        identified_set_exists(boundset, return_one_solution=True)
    """
    solver_setup = set_up_solver(full_bounds_set, check_only)
    rst = solver_setup['rst']
    direc = solver_setup['direc']
    rhs = solver_setup['rhs']

    numsets = rst.shape[1]

    result = _linprog_solve(
        obj=np.zeros(numsets),
        rst=rst,
        direc=direc,
        rhs=rhs,
    )

    if result.status == 0:  # Feasible / optimal
        if return_one_solution:
            setting_names = list(full_bounds_set['univariate_bounds'].keys())
            if check_only is not None:
                setting_names = [
                    s for s in setting_names if s in check_only
                ]
            return dict(zip(setting_names, result.x))
        return True
    else:
        return False


def univariate_bounds_table(full_bounds_set, pin_values=None, **kwargs):
    """
    Univariate Bounds Table.

    For each setting, compute the widest range of effect values consistent
    with *all* univariate and pairwise constraints simultaneously, using
    linear programming.

    This is a univariate summary: it reports the marginal bounds for each
    setting in isolation, ignoring that certain *combinations* across
    settings may be infeasible.  Use :func:`identified_set_exists` with
    ``return_one_solution=True`` or enumerate the joint identified set for
    a fully multivariate picture.

    Parameters
    ----------
    full_bounds_set : dict
        Output of :func:`build_bounds`.
    pin_values : dict, optional
        Mapping of setting names to fixed values.  Pins the effect in
        those settings to the given value before computing bounds for the
        remaining settings.  Values can be:

        * a numeric float — pin to that exact value;
        * ``'lower'`` — pin to the univariate lower bound;
        * ``'upper'`` — pin to the univariate upper bound;
        * ``'original'`` — pin to the point estimate.

        Useful for exploring how fixing one setting's effect constrains
        the rest.
    **kwargs
        Currently unused; accepted for forward compatibility.

    Returns
    -------
    pandas.DataFrame
        One row per setting with columns:

        * ``setting``
        * ``estimate`` — original point estimate.
        * ``lower_bound`` — LP-derived lower bound for the effect.
        * ``upper_bound`` — LP-derived upper bound for the effect.
        * ``lower_nu_bound`` — lower bound on the bias
          ``= estimate - upper_bound``.
        * ``upper_nu_bound`` — upper bound on the bias
          ``= estimate - lower_bound``.
        * ``original_lower_bound`` — input lower plausibility bound
          (before cross-setting tightening).
        * ``original_upper_bound`` — input upper plausibility bound.

    Examples
    --------
    ::

        from hetset import (
            ovb_bounds_by_setting, build_bounds,
            univariate_bounds_table, load_close_college,
        )
        import numpy as np

        df = load_close_college()
        bounds = ovb_bounds_by_setting(
            data=df, setting='blacksouth', treatment='educ',
            model='lwage ~ educ + smsa + married + exper',
            benchmark_covariates=['smsa', 'married', 'exper'], kd=0.1,
        )
        boundmat = np.array([
            [0, 2, 1.5, 0],
            [0, 0, 0, 1.5],
            [0, 0, 0, 2],
            [0, 0, 0, 0],
        ])
        boundset = build_bounds(
            bounds, boundmat,
            settings_order=['Neither', 'Black', 'South', 'Black and South'],
        )
        univariate_bounds_table(boundset)
    """
    setting_names = list(full_bounds_set['univariate_bounds'].keys())

    # Deep copy so we can pin values without mutating the caller's object
    original_bounds_set = copy.deepcopy(full_bounds_set)

    if pin_values is not None:
        for s, val in pin_values.items():
            if isinstance(val, str):
                if val == 'lower':
                    val = full_bounds_set['univariate_bounds'][s]['lower']
                elif val == 'upper':
                    val = full_bounds_set['univariate_bounds'][s]['upper']
                elif val == 'original':
                    val = full_bounds_set['univariate_bounds'][s]['estimate']
                else:
                    raise ValueError(
                        f"pin_values for setting '{s}' must be numeric, "
                        "'lower', 'upper', or 'original'."
                    )
            full_bounds_set['univariate_bounds'][s]['lower'] = float(val)
            full_bounds_set['univariate_bounds'][s]['upper'] = float(val)

    solver_setup = set_up_solver(full_bounds_set, check_only=None)
    rst = solver_setup['rst']
    direc = solver_setup['direc']
    rhs = solver_setup['rhs']

    numsets = len(setting_names)
    rows = []

    for s_idx, s in enumerate(setting_names):
        # ---- Minimise variable s_idx -----------------------------------
        min_obj = np.zeros(numsets)
        min_obj[s_idx] = 1.0
        min_result = _linprog_solve(min_obj, rst, direc, rhs)
        slv_low = min_result.x[s_idx]

        # ---- Maximise variable s_idx (= minimise its negative) ---------
        max_obj = np.zeros(numsets)
        max_obj[s_idx] = -1.0
        max_result = _linprog_solve(max_obj, rst, direc, rhs)
        slv_up = max_result.x[s_idx]

        est = full_bounds_set['univariate_bounds'][s]['estimate']

        rows.append({
            'setting': s,
            'estimate': est,
            'lower_bound': slv_low,
            'upper_bound': slv_up,
            'lower_nu_bound': est - slv_up,
            'upper_nu_bound': est - slv_low,
            'original_lower_bound': (
                original_bounds_set['univariate_bounds'][s]['lower']
            ),
            'original_upper_bound': (
                original_bounds_set['univariate_bounds'][s]['upper']
            ),
        })

    result_df = pd.DataFrame(rows)
    # Recompute to match R's final assignment order exactly
    result_df['lower_bound'] = (
        result_df['estimate'] - result_df['upper_nu_bound']
    )
    result_df['upper_bound'] = (
        result_df['estimate'] - result_df['lower_nu_bound']
    )

    return result_df
